### ISSUE TITLE
False positive when using assertNull and assertSame on same method

### ISSUE THREAD

**Issue by BluePsyduck** [phpstan/phpstan-phpunit#25]
### Summary of a problem or a feature request

When asserting a method result to be null, then changing the internal state of the object, and then asserting the method result to be the same as some object, PHPStan currently fails thinking it will still be null, letting assertSame() always fail.

### Code snippet that reproduces the problem

```php
<?php

declare(strict_types=1);

class Foo {
    /**
     * @var Bar|null
     */
    protected $bar;

    /**
     * @return Bar|null
     */
    public function getBar(): ?Bar
    {
        return $this->bar;
    }

    /**
     * @param Bar|null $bar
     */
    public function setBar(?Bar $bar): void
    {
        $this->bar = $bar;
    }
}

class Bar
{
}

class FooTest extends TestCase
{
    public function testBar(): void
    {
        $foo = new Foo();
        $bar = new Bar();

        $this->assertNull($foo->getBar());
        $foo->setBar($bar);
        $this->assertSame($bar, $foo->getBar()); // <-- This line triggers the error
    }
}
```

### Used configuration

```
parameters:
  level: max

includes:
  - vendor/phpstan/phpstan-phpunit/extension.neon
  - vendor/phpstan/phpstan-phpunit/rules.neon
```
* phpstan/phpstan @ 0.10.2
* phpstan/phpstan-phpunit @ 0.10

### Actual output

```
 ------ --------------------------------------------------------------------------------------------------------------------------------------- 
  Line   FooTest.php                                                                                                                            
 ------ --------------------------------------------------------------------------------------------------------------------------------------- 
  18     Call to method PHPUnit\Framework\Assert::assertSame() with FactorioItemBrowser\ExportData\Bar and null will always evaluate to false.  
 ------ --------------------------------------------------------------------------------------------------------------------------------------- 
                                                                                                                       
 [ERROR] Found 1 error      
```

### Expected output
```
 [OK] No errors   
```

PHPUnit test does run successfully, and so should PHPStan, as all asserts are valid.

**Comment by mhujer** [comment]
It started to happen with 0.10.2 (was fine in 0.10.1).

My workaround is to call `self::assertNotNull($foo->getBar());` to persuade phpstan that it is no longer null.

**Comment by marcospassos** [comment]
Same case here.

**Comment by marcospassos** [comment]
We're facing the same issue here:

```php
class Foo {
    public static function bar() : object {
         return new \stdClass();
    }
}

self::assertSame(Foo::bar(), Foo::bar());
```

```
Call to static method PHPUnit\Framework\Assert::assertSame() with object and object will always evaluate to true.
```

**Comment by fantomas662** [comment]
Still happening in 0.11

**Comment by Jean85** [comment]
Same here. The root issue here seems to stem from the fact that it's deviating from the standard PHPStan behavior, that doesn't lock the return type on a method, only on variables.

**Comment by ste93cry** [comment]
I have the same error on PHPStan 0.10.8 and `phpstan-phpunit` 0.10. I have a code similar to the snippet that reproduces the problem posted in the first comment, however the workaround of @mhujer doesn't work for me

**Comment by ondrejmirtes** [comment]
This will be solved in 0.12.

**Comment by josefbenjac** [comment]
I can confirm, it's resolved

**Comment by ondrejmirtes** [comment]
It's not perfect right now, the example with `setBar()` method works only thanks to the fact it returns `void`.

**Comment by ondrejmirtes** [comment]
It now works if you annotate the `setBar` with `@phpstan-impure`.
